subsystems/DrivetrainBase.java

```diff
@@ -1,51 +1,112 @@
 package org.frc5587.lib.subsystems;
 
+import org.frc5587.lib.advanced.LimitedPoseMap;
 import org.frc5587.lib.pid.FPID;
-import com.revrobotics.CANSparkMax;
+import java.util.Hashtable;
 import com.kauailabs.navx.frc.AHRS;
+import com.ctre.phoenix.motorcontrol.NeutralMode;
 import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
 import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj.kinematics.DifferentialDriveOdometry;
+import edu.wpi.first.wpilibj.kinematics.DifferentialDriveWheelSpeeds;
+import edu.wpi.first.wpilibj.geometry.Pose2d;
+import edu.wpi.first.wpilibj.geometry.Rotation2d;
+import edu.wpi.first.wpilibj.util.Units;
+import edu.wpi.first.wpilibj.Encoder;
 import edu.wpi.first.wpilibj.SpeedControllerGroup;
+import edu.wpi.first.wpilibj.Timer;
 import edu.wpi.first.wpilibj.controller.PIDController;
 import edu.wpi.first.wpilibj2.command.PIDSubsystem;
 
 public abstract class DrivetrainBase extends PIDSubsystem {
-    // create leader and follower motors for the drivetrain
+    // create leader and follower motors for the drivetrain, as well as encoders that will attach to the leaders.
     protected WPI_TalonFX leftLeader, rightLeader, leftFollower, rightFollower;
+    protected Encoder leftEncoder, rightEncoder;
 
     // group the leader and follower motors so they can be controlled at the same time
     protected SpeedControllerGroup leftGroup, rightGroup;
 
     // make the speed controller groups into one drivetrain object
     protected DifferentialDrive differentialDrive;
 
-    // create variables needed for PID. these will all be changed by the subclass.
+    // create a hashtable to look up constants
+    protected Hashtable<String, Object> constants;
```

Comment:
If you're going to use a hash map instead of an object, document which keys are needed and what they're supposed to be